Add explicit exceptions during command calls / handling

hikinggrass opened the pull request:
Allow exceptions thrown in cmd handlers to be propagated to the calling module.

To enable this you can set the following in a config file:
```yaml
settings:
  forward_exceptions: true
````

This allows the caller to handle certain error conditions more gracefully (and could be extended for shutdown handling during cmd callc in the future as well)

A module could handle these new exceptions in the following way:

```cpp
try {
    req->call_something();
} catch (const Everest::HandlerException& e) {
    EVLOG_error << "got a HandlerException: " << e.what();
} catch (const Everest::CmdTimeout& e) {
    EVLOG_error << "got a CmdTimeout: " << e.what();
} catch (const Everest::CmdError& e) {
    EVLOG_error << "got a CmdError: " << e.what();
}
```

`HandlerExceptions` and `CmdTimeout` inherit from `CmdError` so a caller could also just catch this.

hikinggrass committed:
WIP: exceptions during command calls / handling

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
WIP: Use ErrorType in CmdResult

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/cmd-exceptions

# Conflicts:
#	lib/everest.cpp

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
cleanup

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Rename EverestCmdError to CmdError

Handle cmd timeout with new cmd error mechanism

Add HandlerException and CmdTimeout exceptions that modules can specifically handle

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge branch 'main' into feature/cmd-exceptions

hikinggrass committed:
reorder error type to string conversions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Throw CmdError in the unlikely event that a provided cmd does not contain an id

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Reorder to/from json ErrorMessage conversions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass commented on include/framework/everest.hpp:
```diff
@@ -34,6 +34,49 @@ using TelemetryEntry = std::variant<std::string, const char*, bool, int32_t, uin
 using TelemetryMap = std::map<std::string, TelemetryEntry>;
 using UnsubscribeToken = std::function<void()>;
 
+enum class ErrorType {
```
`ErrorType` might be a bit vague and could lead to confusions with the "error framework"

hikinggrass replied:
Maybe call this `CallCmdEvent` or `CmdEvent`

hikinggrass commented on include/framework/everest.hpp:
```diff
@@ -34,6 +34,49 @@ using TelemetryEntry = std::variant<std::string, const char*, bool, int32_t, uin
+    HandlerException,
+    Timeout,
+    Shutdown,
+    Unknown
```
remove Unknown

hikinggrass commented on include/framework/everest.hpp:
```diff
@@ -34,6 +34,49 @@ using TelemetryEntry = std::variant<std::string, const char*, bool, int32_t, uin
+};
+
+struct ErrorMessage {
+    ErrorType type;
```
event

hikinggrass commented on include/framework/everest.hpp:
```diff
@@ -34,6 +34,49 @@ using TelemetryEntry = std::variant<std::string, const char*, bool, int32_t, uin
+    Unknown
+};
+
+struct ErrorMessage {
```
`CmdResultError`

hikinggrass commented on include/framework/everest.hpp:
```diff
@@ -34,6 +34,49 @@ using TelemetryEntry = std::variant<std::string, const char*, bool, int32_t, uin
+public:
+    using CmdError::CmdError;
+};
+
```
Maybe already define exceptions for the other `CallCmdEvent`s

hikinggrass commented on lib/everest.cpp:
```diff
@@ -863,19 +895,26 @@ void Everest::provide_cmd(const std::string& impl_id, const std::string& cmd_nam
+            EVLOG_verbose << fmt::format("Exception during handling of: {}->{}({}): {}",
+                                         this->config.printable_identifier(this->module_id, impl_id), cmd_name,
+                                         fmt::join(arg_names, ","), e.what());
+            error = ErrorMessage{ErrorType::HandlerException, e.what()};
```
re-throw after result has been sent to caller so the handler still crashes

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/cmd-exceptions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Renamings

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add exceptions for all CmdEvents

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
re-throw caught handler exception after caller has been notified via mqtt

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Also handle unknown exceptions during handler calls

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove "Unknown" CmdEvent

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move new exceptions into types.hpp

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
forward declare CmdEvent, CmdResultError and CmdResult

move conversions into its own file

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Rename Events that describe failures with "Failure" postfix

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass commented on lib/everest.cpp:
```diff
@@ -879,16 +923,30 @@ void Everest::provide_cmd(const std::string& impl_id, const std::string& cmd_nam
+                    throw std::runtime_error("Unknown exception during cmd handling");
+                }
+            }
+        }
```
TODO: should this be configurable? default = rethrow (which is very similar to the current behavior) but can be turned off?

Pietfried replied:
For now I would just rethrow it and possibly make it configurable in a subsequent PR after we have discussed and analyzed potential implications of that

hikinggrass replied:
default is now still the current behavior, but the new behavior can be turned on by setting the following in a config:
```yaml
settings:
  forward_exceptions: true
````

hikinggrass committed:
Rename type to event

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add a few simple tests for the new conversions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
"type" is now known as "event"

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

dorezyuk commented on include/utils/types.hpp:
```diff
@@ -187,9 +187,59 @@ NLOHMANN_JSON_NAMESPACE_END
+    SchemaValidationFailed,
+    HandlerException,
+    Timeout,
+    Shutdown
```
should we add also a notyetready?

hikinggrass replied:
added NotReady enum

Pietfried commented on lib/conversions.cpp:
```diff
@@ -0,0 +1,61 @@
+    switch (cmd_event) {
+    case CmdEvent::MessageParsingFailed:
+        return CMD_EVENT_MESSAGE_PARSING_FAILED;
+        break;
```
Can be removed here and following
```suggestion
```

Pietfried commented on lib/everest.cpp:
```diff
@@ -401,18 +411,32 @@ json Everest::call_cmd(const Requirement& req, const std::string& cmd_name, json
+        throw CmdError("Command did not return result");
+    } else {
+        return result.result.value();
+    }
```
nit: reads a little better
```suggestion
if (result.error.has_value()) {
        const auto& error = result.error.value();
        const std::string err_msg = fmt::format("{}", error.msg);
    
        switch (error.event) {
            case CmdEvent::HandlerException:
                throw HandlerException(err_msg);
            case CmdEvent::Timeout:
                throw CmdTimeout(err_msg);
            default:
                throw CmdError(fmt::format("{}: {}", conversions::cmd_event_to_string(error.event), error.msg));
        }
    }
    
    if (!result.result.has_value()) {
        throw CmdError("Command did not return result");
    }
    
    return result.result.value();
```

Pietfried commented on lib/everest.cpp:
```diff
@@ -851,19 +883,31 @@ void Everest::provide_cmd(const std::string& impl_id, const std::string& cmd_nam
+            EVLOG_error << fmt::format("Unknown exception during handling of: {}->{}({})",
+                                       this->config.printable_identifier(this->module_id, impl_id), cmd_name,
+                                       fmt::join(arg_names, ","));
+            error = CmdResultError{CmdEvent::HandlerException, "Unknown exception"};
```
couldnt we use `std::current_exception()` here as well?

hikinggrass replied:
added usage of `std::current_exception()`

dorezyuk commented on include/utils/types.hpp:
```diff
@@ -187,9 +187,59 @@ NLOHMANN_JSON_NAMESPACE_END
+
+struct CmdResult {
+    std::optional<json> result;
+    std::optional<CmdResultError> error;
```
to my understanding they are mutually exclusive - can we express that through the type system, e.x a variant?

hikinggrass replied:
A variant here might a bit tricky at the moment due to `CmdResultError` being convertible to `json` and result being a `json` type. But let's keep this in mind during the upcoming refactorings in this part of the code

hikinggrass committed:
Add "NotReady" CmdEvent and explanations for the different events

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Remove redundant "break"

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Restructure exception throwing during call_cmd based on review comments

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use std::current_exception for every caught exception

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass commented on lib/everest.cpp:
```diff
@@ -401,18 +411,37 @@ json Everest::call_cmd(const Requirement& req, const std::string& cmd_name, json
+            throw HandlerException(error_message);
+        case CmdEvent::Timeout:
+            throw CmdTimeout(error_message);
+        default:
```
TODO: use other defined exceptions here

Pietfried replied:
This doesnt look like it is resolved?

What about?
MessageParsingFailed
SchemaValidationFailed
Shutdown
NotReady

hikinggrass replied:
Good catch, I have added the missing exception throws

hikinggrass commented on include/utils/types.hpp:
```diff
@@ -187,9 +187,66 @@ NLOHMANN_JSON_NAMESPACE_END
     using std::runtime_error::runtime_error;
 };
+
+class CmdError : public std::runtime_error {
```
move this into a separate file

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/cmd-exceptions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Move exceptions into their own header

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/cmd-exceptions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
clang-format

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Add setting to forward handler exceptions

Set to false by default which will re-throw these exceptions like the current behavior
Set to true a handler exception will be caught, forwarded to the caller and not re-thrown in the handler itself

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Passing forward exceptions setting for C++ modules

Logging message if exception forwarding is enabled

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Use forward exceptions parameter in Python, Rust and JavaScript bindings as well

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/cmd-exceptions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Merge remote-tracking branch 'origin/main' into feature/cmd-exceptions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

hikinggrass committed:
Set FORWARD_EXCEPTION to a default of 0 in a migrated database

This is needed because the column is marked as NOT NULL and therefore always needs a value

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>

Pietfried reviewed (changes requested):
lgtm, except one comment in line

hikinggrass committed:
Add missing exceptions

Signed-off-by: Kai-Uwe Hermann <kai-uwe.hermann@pionix.de>
